Differential fuzzing

aeyno opened the pull request:
This PR adds support for differential fuzzing using Puffin.
Currently, only TLS is supported with wolfSSL and OpenSSL CPUTs.

aeyno committed:
basis for differential fuzzing

[differential fuzzing] fixing tracerunner change for rebase

adding cli syntax to start differential fuzzing campaign

[differential fuzzing] cleaning

aeyno committed:
[differential fuzzing] Update FuzzingTarget to use PutDescriptor instead of String

aeyno committed:
[differential fuzzing] Comparing trace execution status

aeyno committed:
[differential fuzzing] Add whitelist/blacklist to filter knowledges to compare

aeyno committed:
[differential fuzzing] adding TraceDifference enum

aeyno committed:
[differential fuzzing] Comparing knowledges with Compare crate

aeyno committed:
[differential fuzzing] Extend shortest knowledge store before comparison

aeyno committed:
[differential fuzzing] Showing source in knowledge comparison

aeyno committed:
[differential fuzzing] removing some comparable_ignore for tls

aeyno committed:
[differential fuzzing] Comparing claims

[differential fuzzing] Improving comparisons of knowledges and claims

[differential fuzzing] Improving comparisons of knowledges and claims

[differential puffin] improving claims

aeyno committed:
[differential fuzzing] Infrastructure to decrypt knowledges

aeyno committed:
[differential fuzzing] override put config to use a restricted set of ciphers

[differential fuzzing] setting default groups for tls

fix setting tls ciphers

aeyno committed:
[differential fuzzing] adding --differential flag to seed generation to override agent descriptors

aeyno committed:
[differential fuzzing][temp] removing comparison for all claims that are not in the openssl/wolfssl claim intersection

aeyno committed:
[differential fuzzing] sorting claims by agent and deduplicating similar adjacent claims

[differential-fuzzing][tlspuffin] ignore step in tls claim

aeyno committed:
[boringssl] add Finished security claim for BoringSSL

fix add early and handhake secrets to boringssl finished claim

aeyno committed:
[puffin] display PUT errors when using execute

aeyno committed:
[differential fuzzing] comparing execution step when both PUT fail

aeyno committed:
[differential fuzzing] improve the structure of the TraceDifference object

aeyno committed:
[differential fuzzing] allow differential execute to export differences as JSON

aeyno committed:
[differential fuzzing] improve error filtering

aeyno committed:
[differential fuzzing][tls] add decryption of handshake

[tlspuffin] removing already defined function symbol

aeyno committed:
[tlspuffin] change TLS 1.2 differential fuzzing cipher list to fix broken seed

aeyno committed:
[differential fuzzing][tls] decryption of client finished message

aeyno committed:
[differential-fuzzing][tlspuffin] ignore ECPointFormatList in TLS 1.2 handshake messages to prevent false positives

aeyno committed:
[differential fuzzing][tlspuffin] add Flight TlsQueryMatcher to improve knowledge decryption

aeyno committed:
[differential fuzzing] add security claims violation as differences

aeyno committed:
[differential fuzzing] add source to knowledge differences

aeyno committed:
[differential fuzzing] changing the way we uniformize Agent config

aeyno committed:
[differential fuzzing] allow to show decrypted term with display-execute

aeyno committed:
[differential fuzzing] add post execution diff filter + ignore some fields

aeyno committed:
[differential fuzzing] fix comparing PUT status

aeyno committed:
[puffin][display-execute] allow exporting evaluated terms as raw hex

aeyno committed:
[differential fuzzing][tlspuffin] sort HelloRetryRequest extensions to avoid false positives

aeyno committed:
[differential fuzzing] allow disabling security oracle checks when calling execute

[differential fuzzing] fix call to execute after changing signature

[differential fuzzing] changing flag to disable security oracle in display-execute

[differential fuzzing] fix typo

aeyno committed:
[differential fuzzing] add differential-experiment cli command

aeyno committed:
[differential fuzzing] add feature flags for ablation study

aeyno committed:
[differential fuzzing] adding documentation + evaluation scripts + triaging tools

nataelbaffou commented on puffin/src/claims.rs:
```diff
@@ -94,6 +108,91 @@ impl<C: Claim> ClaimList<C> {
         self.claims.push(claim);
     }
+
+    pub fn compare(&self, other: &Self) -> Result<(), Vec<TraceDifference>> {
+        let blacklist = <C::PT as ProtocolTypes>::differential_fuzzing_claims_blacklist();
```
Maybe add some comments to explain the logic behind the processing of the claims, like explained if the conf (dedupping, ordering, ...)

aeyno replied:
done

nataelbaffou commented on puffin/src/cli.rs:
```diff
@@ -80,7 +81,9 @@ where
                 .arg(arg!(-k --show_knowledges "Show the knowledges gathered at each output step").value_parser(value_parser!(bool)))
                 .arg(arg!(-r --show_raw "Show the computed term as raw hex (eg. for use with netcat)").value_parser(value_parser!(bool)))
-                .arg(arg!(-j --json "Export trace execution as JSON").value_parser(value_parser!(bool))),
+                .arg(arg!(-p --differential_post_computations "Evaluate the post execution terms ued in differential fuzzing").value_parser(value_parser!(bool)))
```
typo
```suggestion
                .arg(arg!(-p --differential_post_computations "Evaluate the post execution terms used in differential fuzzing").value_parser(value_parser!(bool)))
```

aeyno replied:
Done

nataelbaffou commented on puffin/src/differential.rs:
```diff
@@ -0,0 +1,190 @@
+use crate::error::Error;
+use crate::trace::Source;
+
+/// A difference between to exection in differential fuzzing
```
typo
```suggestion
/// A difference between to execution in differential fuzzing
```

aeyno replied:
Done

nataelbaffou commented on puffin/src/cli.rs:
```diff
@@ -218,19 +237,22 @@ where
                 (false, env::current_dir().unwrap())
             }
         };
-    if is_experiment {
```
is_experiment variable is not longer is use, you can remove it and all respective returned values (true, false) from the if-else statement

aeyno replied:
done

nataelbaffou commented on puffin/src/protocol.rs:
```diff
@@ -24,6 +26,74 @@ impl<T: 'static> AsAny for T {
+    fn boxed(&self) -> Box<dyn EvaluatedTerm<PT>> {
+        Box::new(self.clone())
+    }
+}
```
It seems unused as the "boxed" method was already is "EvaluatedTerm" trait. Did you want to replace it ?

aeyno replied:
Done

nataelbaffou commented on puffin/src/trace.rs:
```diff
@@ -643,19 +798,27 @@ impl<PT: ProtocolTypes> TraceExecution<PT> {
                     }
                 }
                 Some(step_knowledges)
+                // TODO : Bump rust version to 1.87 to use `extract_if` instead of mem:swap code
+                // Some(
+                //     ctx.knowledge_store
+                //         .raw_knowledge
+                //         .extract_if(.., |k| k.step == Some(StepNumber::new(trace_number, idx)))
+                //         .map(|k| k.data)
+                //         .collect(),
+                // )
```
@LCBH should we add an issue or should it be fixed in the "update_rustc_cargo_versions" branch ?

aeyno replied:
duplicate

LCBH replied:
This will be doable now.

nataelbaffou commented on puffin/src/trace.rs:
```diff
@@ -669,13 +832,40 @@ impl<PT: ProtocolTypes> TraceExecution<PT> {
+                }
+            }
+            Some(knowledges)
+            // TODO : Bump rust version to 1.87 to use `extract_if` instead of mem:swap code
+            // Some(
+            //     ctx.knowledge_store
+            //         .raw_knowledge
+            //         .extract_if(.., |k| k.step == None)
+            //         .map(|k| (k.source, k.data))
+            //         .collect(),
+            // )
```
(COPY) @LCBH should we add an issue or should it be fixed in the "update_rustc_cargo_versions" branch ?

aeyno replied:
As discussed IRL, we will do another PR to replace the code when the rust version will be updated

nataelbaffou commented on sshpuffin/Cargo.toml:
```diff
@@ -23,6 +23,7 @@ puffin-build = { path = "../puffin-build" }
 extractable-macro = { path = "../extractable-macro" }
 log = { workspace = true }
 serde = { workspace = true, features = ["derive"] }
+comparable = { workspace = true }
```
You need the "derive" feature and you get it from the puffin/Cargo.toml. But is it a better practive to also include it here ?

aeyno replied:
done

nataelbaffou commented on tlspuffin/Cargo.toml:
```diff
@@ -354,6 +359,7 @@ hex = { workspace = true }
 serde = { workspace = true, features = ["derive"] }
 smallvec = { workspace = true }
 once_cell = { workspace = true }
+comparable = { workspace = true }
```
(COPY) You need the "derive" feature and you get it from the puffin/Cargo.toml. But is it a better practive to also include it here ?

aeyno replied:
Done

nataelbaffou reviewed (commented):
# Fixes
- sshpuffin/src/ssh/seeds.rs:183 -> missing 3rd parameter in execute method
  - We don't run the sshpuffin tests in the CI ?
- extractable-macro/tests/tests.rs -> errors with trait implementations

# Other notes
- Should some tests for the differential fuzzing be added ? (I might have missed them)
- Lots of scripts in evaluation-ddyf, maybe add an explanation at the beginning of each one to explain a bit more what it does or an MD file to put all those explanations ? (Kinda like the readme that is the artifact of the publication paper)

Maybe I am being too nitpicky, the work is really great and the code clean.

nataelbaffou commented on puffin/src/cli.rs:
```diff
@@ -93,7 +96,23 @@ where
+                .arg(arg!(<second_target> "Second PUT"))
+                .arg(arg!(<input> "Input trace"))
+                .arg(arg!(-j --json "Export differences as JSON").value_parser(value_parser!(bool))),
+            Command::new("differential")
```
Maybe change the command to "quick-differential" for consistency

aeyno replied:
Technically it is not an experiment (objectives are stored in the `objectives` folder) and you can't provide any experiment title like with quick-experiment. This command is closer to running a differential fuzzing campaign without using the fuzzer without any command.

nataelbaffou commented on puffin/src/cli.rs:
Add "differential" or "quick-differential" here for consistency

aeyno replied:
duplicate

aeyno committed:
[differential fuzzing][evaluation] adding script to count ignored tls fields

aeyno committed:
[differential fuzzing] cleaning + small fixes + fixing typos

aeyno committed:
[differential fuzzing] fixing extractable-macro

LCBH commented on puffin/src/fuzzer/harness.rs:
```diff
@@ -35,9 +38,13 @@ pub fn harness<PB: ProtocolBehavior + 'static>(
     }
     // Execute the trace
-    let runner = Runner::new(put_registry.clone(), Spawner::new(put_registry.clone()));
+    let agent = input.descriptors[0].name;
```
Protect this in case descriptors is empty.

aeyno replied:
done

LCBH commented on puffin/src/fuzzer/harness.rs:
```diff
@@ -35,9 +38,13 @@ pub fn harness<PB: ProtocolBehavior + 'static>(
+    let agent = input.descriptors[0].name;
+    let runner = Runner::new(
+        put_registry.clone(),
+        Spawner::new(put_registry.clone()).with_mapping(&[(agent, put_descriptor.clone())]),
```
Why don't we map all agents from the trace? Here, we only do so for the first in the list.

aeyno replied:
done

LCBH commented on puffin/src/fuzzer/harness.rs:
```diff
@@ -57,6 +64,88 @@ pub fn harness<PB: ProtocolBehavior + 'static>(
+    second_put: &PutDescriptor,
+    input: &Trace<PB::ProtocolTypes>,
+) -> ExitKind {
+    let agent = input.descriptors[0].name;
```
Same here

aeyno replied:
done

LCBH commented on puffin/src/protocol.rs:
```diff
@@ -178,6 +237,32 @@ pub trait ProtocolTypes:
+    ) -> Vec<crate::algebra::Term<Self>>;
+
+    /// Replace trace's agent protocol config with parameters
+    /// that unifomize the behavior of the PUTs
```
typo

LCBH commented on puffin/src/protocol.rs:
```diff
@@ -178,6 +237,32 @@ pub trait ProtocolTypes:
+    /// Terms to evaluate and compare during differential fuzzing
+    /// eg. decryption recipes
+    /// Terms to evaluate and compare during differential fuzzing
+    /// eg. decryption recipes
```
typo

aeyno replied:
done

LCBH commented on puffin/src/protocol.rs:
```diff
@@ -293,3 +397,29 @@ macro_rules! atom_extract_knowledge {
+            ) {
+                match other.as_any().downcast_ref::<$extract_type>() {
+                    Some(_) => {
+                        todo!("Comparable for {}", other.type_name());
```
Do we really want to panic here? Don't we want log::err instead?

aeyno replied:
This macro was not used anymore so I removed it

LCBH commented on puffin/src/protocol.rs:
```diff
@@ -178,6 +237,32 @@ pub trait ProtocolTypes:
+    /// that unifomize the behavior of the PUTs
+    fn differential_fuzzing_uniformise_put_config(trace: Trace<Self>) -> Trace<Self>;
+
+    /// Check wether a difference should be kept
```
typo

aeyno replied:
Done

LCBH commented on puffin/Cargo.toml:
```diff
@@ -25,6 +25,11 @@ with-min-scheduler = []
 
 introspection = ["libafl/introspection"]
 
+ddyf-disable-status = []
```
Add a one-liner comment to explain this is for ablation study

aeyno replied:
Done

LCBH commented on puffin/src/claims.rs:
```diff
@@ -94,6 +108,96 @@ impl<C: Claim> ClaimList<C> {
+            let s = self_claims_filtered.get(k).unwrap_or(&empty);
+            let o = other_claims_filtered.get(k).unwrap_or(&empty);
+
+            for i in 0..usize::max(s.len(), o.len()) {
```
Can't we simply apply Comparable on the two lists? (hmm not sure actually since we treat differently the case different content from the case missing content...)

aeyno replied:
We could but this increases the readability of the differences: else we would only have a big data structure with all the differences (inner diff and different types) for all the claims.

LCBH commented on puffin/src/codec.rs:
```diff
@@ -417,3 +417,10 @@ impl<T: Codec + VecCodecWoSize> Codec for Vec<T> {
+pub fn compare_encoding<X: Codec, Y: Codec>(x: &X, y: &Y) -> std::cmp::Ordering {
+    match Codec::get_encoding(x) <= Codec::get_encoding(y) {
+        true => std::cmp::Ordering::Less,
+        false => std::cmp::Ordering::Greater,
```
Why no Equal case? Seems to violate the Ord contract

aeyno replied:
Changed to support equal case

LCBH commented on puffin/src/put_registry.rs:
```diff
@@ -46,6 +46,10 @@ impl<PB> PutRegistry<PB> {
     }
+
+    pub fn default_put_descriptor(&self) -> PutDescriptor {
+        PutDescriptor::new(self.default_put_name(), PutOptions::empty())
```
Use `self.default_put.options` rather. than `PutOptions::empty()`?!

aeyno replied:
done

LCBH commented on puffin/src/differential.rs:
```diff
@@ -0,0 +1,190 @@
+        first_source: Source,
+        second_source: Source,
+    },
+    /// The two knowledges have the same type but there internal values are different
```
typo

aeyno replied:
done

LCBH commented on puffin/src/cli.rs:
```diff
@@ -217,20 +236,29 @@ where
         };
-    if is_experiment {
-        log::info!("Running in experiment mode");
-        config.is_experiment = is_experiment;
```
We never set config.is_experiment to true now, which is needed for correct logging (see where it was used). Was this intentional?

aeyno replied:
This looks like an error during a rebase. I put it back.

LCBH commented on puffin/src/cli.rs:
```diff
@@ -217,20 +236,29 @@ where
                     "Experiment already exists. Consider creating a new experiment."
                 );
-                (true, experiment_path)
+                experiment_path
```
Since the two branches have the same code, merge them.

aeyno replied:
done

LCBH commented on puffin/src/cli.rs:
```diff
@@ -434,15 +484,79 @@ where
+            return ExitCode::FAILURE;
+        }
+
+        let agent = trace.descriptors[0].name;
```
Better deal with cases where the list is empty.

LCBH replied:
Another occurence of this in this file

aeyno replied:
done

LCBH commented on puffin/src/cli.rs:
```diff
@@ -434,15 +484,79 @@ where
+        let runner = DifferentialRunner::new(
+            put_registry.clone(),
+            Spawner::new(put_registry.clone())
+                .with_mapping(&[(agent, PutDescriptor::new(first_put, PutOptions::empty()))]),
```
Why only the first agent of the trace is set?

aeyno replied:
done

LCBH commented on puffin/src/cli.rs:
```diff
@@ -434,15 +484,79 @@ where
+                .with_mapping(&[(agent, PutDescriptor::new(second_put, PutOptions::empty()))]),
+        );
+
+        return match runner.execute(trace, &mut 0, false) {
```
Is the `false` intended here?

aeyno replied:
I added `--enable-security-oracle` flag to enable the security oracle which is disabled by default in differential mode.

LCBH commented on puffin/src/execution.rs:
```diff
@@ -8,27 +9,39 @@ use nix::sys::wait::{waitpid, WaitPidFlag};
+            trace,
+            ConfigTrace::default(),
+            executed_until,
+            check_security_violation,
```
Wouldn't be cleaner to includer this into ConfigTrace instead?

LCBH replied:
Hmm, would need to add this parameter to `execute` as well then. To be discussed.

aeyno replied:
Done.

LCBH commented on puffin/src/execution.rs:
```diff
@@ -138,6 +164,179 @@ impl<T: TraceRunner + Clone> TraceRunner for &ForkedRunner<T> {
+    type PB = PB;
+    type R = TraceContext<Self::PB>;
+
+    /// Execute `trace` on the two PUT of `self`, return `Ok` if both execution exhibit no
```
typos

aeyno replied:
done

LCBH commented on puffin/src/execution.rs:
```diff
@@ -138,6 +164,179 @@ impl<T: TraceRunner + Clone> TraceRunner for &ForkedRunner<T> {
+        *executed_until = usize::max(first_ctx.executed_until, second_ctx.executed_until);
+
+        //check if we have security claim violation (if security claims are enabled)
+        let mut diff = match (&first_trace_status, &second_trace_status) {
```
Is it intended to check this only if there is no status difference? I feel that security claim violations are "more severe" and less prone to false positives....

aeyno replied:
We just abort the comparison as soon as we detect a status difference for performance reasons , but this kind of differences should be catched during the trace analysis.

aeyno replied:
TODO: make a security violation obvious to be sure to correctly triage it

aeyno replied:
Done.

LCBH commented on puffin/src/execution.rs:
```diff
@@ -138,6 +164,179 @@ impl<T: TraceRunner + Clone> TraceRunner for &ForkedRunner<T> {
+            return Err(Error::Difference(diff));
+        }
+
+        Ok(first_ctx)
```
You never want to access both contexts? I reckon that you already checked that they are both "the same" but this still allows some differences that are filtered out.

aeyno replied:
Changed to return both

LCBH commented on puffin/src/execution.rs:
```diff
@@ -138,6 +164,179 @@ impl<T: TraceRunner + Clone> TraceRunner for &ForkedRunner<T> {
+        self.execute_config(
+            trace,
+            ConfigTrace {
+                with_bit_level: false,
```
Why not letting the user choose this. It would be nice to do diff fuzzing with havoc mutations enabled.

LCBH replied:
See how it is done for the other implem of `execute`.

aeyno replied:
Changed.

LCBH commented on puffin/src/trace.rs:
```diff
@@ -258,6 +261,92 @@ impl<PT: ProtocolTypes> KnowledgeStore<PT> {
+    }
+}
+
+/// Should a specific knowledge be filtered out
```
"should be kept (not filtered out)" right?

aeyno replied:
Indeed. Changed.

LCBH commented on puffin/src/trace.rs:
```diff
@@ -258,6 +261,92 @@ impl<PT: ProtocolTypes> KnowledgeStore<PT> {
+    }
+
+    if let Some(list) = whitelist {
+        if !list.iter().any(|x| x == &knowledge.data.type_id()) {
```
Hmmm, I am confused. If it is in the white list --> it is always kept. If not, it is kept only if it is not in the blacklist. No?

aeyno replied:
I don't think that the blacklist is really useful, I will probably remove it

LCBH commented on puffin/src/trace.rs:
```diff
@@ -643,19 +798,27 @@ impl<PT: ProtocolTypes> TraceExecution<PT> {
+                    ctx.claims
+                        .deref_borrow()
+                        .iter()
+                        .filter(|c| c.get_step() == Some(StepNumber::new(trace_number, idx)))
```
This whole block and this filtering needs documentation

LCBH replied:
Actually yhe whole function would need better documentation.

aeyno replied:
Done

LCBH commented on puffin/src/trace.rs:
```diff
@@ -669,13 +832,40 @@ impl<PT: ProtocolTypes> TraceExecution<PT> {
             executed_until: ctx.executed_until,
             steps,
             // right now we exclude prior traces
             prior_traces: Vec::new(),
```
Why not just putting here the prior traces from the trace?

aeyno replied:
Discussed offline. This requires some changes to the way the knowledges are stored to correctly map them to the correct prior trace. I opened an issue: https://github.com/tlspuffin/tlspuffin/issues/476

LCBH commented on puffin/src/protocol.rs:
```diff
@@ -293,3 +397,29 @@ macro_rules! atom_extract_knowledge {
 }
+
+#[macro_export]
+macro_rules! dummy_compare {
```
Is this used somewhere (I cannot find any occurence)

aeyno replied:
Indeed it was not used anymore so I removed it

LCBH commented on puffin/src/protocol.rs:
```diff
@@ -293,3 +397,29 @@ macro_rules! atom_extract_knowledge {
+macro_rules! dummy_compare {
+    ($protocol_type:ty, $extract_type:ty) => {
+        impl $crate::protocol::CompareKnowledge<$protocol_type> for $extract_type {
+            fn find_differences(
```
Wrong function signature!

aeyno replied:
This macro has been removed

LCBH commented on puffin/src/trace.rs:
```diff
@@ -258,6 +261,92 @@ impl<PT: ProtocolTypes> KnowledgeStore<PT> {
+        log::trace!("Comparing knowledge stores");
+        let _ = std::iter::zip(first_store, second_store)
+            .enumerate()
+            .map(|(idx, (x, y))| {
```
Rather uses `.for_each()`?

aeyno replied:
Done

LCBH commented on tlspuffin/harness/openssl/src/put.c:
```diff
@@ -443,10 +443,6 @@ static RESULT get_result(AGENT agent, int retcode, bool allow_would_block)
     case SSL_ERROR_WANT_READ:
     case SSL_ERROR_WANT_WRITE:
         error_type = strdup("IO_WOULD_BLOCK");
-        if (!allow_would_block)
```
This followed quite a lot of discussions and was not trivial at all. It would be nice to document a bit the rationale. Ideally, add a documentation at the start of the case distinctions with the rationale.

aeyno replied:
Done

nataelbaffou commented on puffin/src/cli.rs:
```diff
@@ -217,20 +236,29 @@ where
-    if is_experiment {
-        log::info!("Running in experiment mode");
-        config.is_experiment = is_experiment;
-        config.corpus_dir = base_directory.join("corpus");
```
Il faut garder ces configs pour tout avoir dans l'experiment en cours et non à la racine du projet

aeyno replied:
done

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -172,9 +173,11 @@ impl From<[u8; 32]> for Random {
-#[derive(Copy, Clone)]
+#[derive(Copy, Clone, Comparable)]
 pub struct SessionID {
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
Since we already ignore SessionIDs in CH and SH, I removed the ignore here

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -357,12 +361,13 @@ impl ServerNamePayload {
     #[extractable_ignore]
     pub typ: ServerNameType,
     #[extractable_ignore]
+    #[comparable_ignore]
```
[Compare] Just for the discussion: do we agree that we could theoretically make sure to configure agents so that this field should be aligned? (If yes, just add a comment saying this)

aeyno replied:
Yes probably, I added the comment.

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -554,9 +562,11 @@ impl codec::Codec for PresharedKeyOffer {
 pub struct OCSPCertificateStatusRequest {
+    #[comparable_ignore]
     pub responder_ids: ResponderIDs,
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
I tried to remove the ignores and I don't have differences on the seed so I am not sure why I ignored them

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -630,10 +640,10 @@ declare_u8_vec!(PSKKeyExchangeModes, PSKKeyExchangeMode);
 #[extractable(TLSProtocolTypes)]
 pub enum ClientExtension {
-    ECPointFormats(ECPointFormatList),
+    ECPointFormats(#[comparable_ignore] ECPointFormatList),
```
[Compare] Just for the discussion: do we agree that we could theoretically make sure to configure agents so that this field should be aligned? (If yes, just add a comment saying this)

aeyno replied:
Done

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -813,16 +823,16 @@ impl ClientExtension {
+#[derive(Clone, Debug, PartialEq, Comparable)]
 pub enum ServerExtension {
-    ECPointFormats(ECPointFormatList),
+    ECPointFormats(#[comparable_ignore] ECPointFormatList),
```
[Compare] Just for the discussion: do we agree that we could theoretically make sure to configure agents so that this field should be aligned? (If yes, just add a comment saying this)

aeyno replied:
Done

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -951,14 +961,25 @@ declare_u16_vec_empty!(ClientExtensions, ClientExtension);
 #[extractable(TLSProtocolTypes)]
 pub struct ClientHelloPayload {
+    #[comparable_synthetic {
+        let sorted_cipher_suites = |x: &Self| -> std::collections::HashMap<CipherSuite, ()> { let ciphers : std::collections::HashMap<CipherSuite, ()>= x.cipher_suites.0.clone().into_iter().map(|c| (c,())).collect(); ciphers };
```
[compare] why removing duplicates? Don't we want to compare as multisets? (as for extensions)

aeyno replied:
Ideed, I changed that

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -1748,12 +1792,13 @@ impl codec::Codec for ClientECDHParams {
 #[extractable(TLSProtocolTypes)]
 pub struct ServerECDHParams {
     #[extractable_ignore]
     pub curve_params: ECParameters,
```
[Compare] Just for the discussion: do we agree that we could theoretically make sure to configure agents so that this field should be aligned? (If yes, just add a comment saying this)

aeyno replied:
Done

LCBH commented on tlspuffin/src/tls/rustls/msgs/handshake.rs:
```diff
@@ -2059,10 +2104,12 @@ impl CertificateRequestPayloadTLS13 {
+#[derive(Debug, Clone, Extractable, Comparable)]
 #[extractable(TLSProtocolTypes)]
 pub struct NewSessionTicketPayload {
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
Yes, the lifetime and ticket payloads are PUT dependant. I don't think it is possible to change this value for OpenSSL

LCBH commented on tlspuffin/src/tls/rustls/msgs/message.rs:
```diff
@@ -53,6 +54,24 @@ impl codec::CodecP for MessagePayload {
     }
 }
+
+impl PartialEq for MessagePayload {
```
Why do we need a custom PartialEq implem here?

aeyno replied:
I think that is to avoid having to derive `PartialEq` on all the underlying types (basically all rustls types)

LCBH commented on tlspuffin/src/tls/rustls/hash_hs.rs:
```diff
@@ -75,9 +76,10 @@ impl HandshakeHashBuffer {
+#[derive(Clone, Comparable)]
 pub struct HandshakeHash {
     /// None before we know what hash function we're using
+    #[comparable_ignore]
```
[Compare] Wait: why don't we have to ignore the other fields (like `override_buffer`)?

aeyno replied:
The reason why this field is ignored is that now `digest::Context` doesn't have to implement Comparable.

LCBH commented on tlspuffin/src/tls/fn_transcript.rs:
```diff
@@ -36,3 +38,35 @@ fn _fn_transcript<T: Transcript>(claim: &T) -> Result<HandshakeHash, FnError> {
     Ok(hash)
 }
+
+pub fn fn_finished_get_cipher(claim: &Finished) -> Result<CipherSuite, FnError> {
```
Please add somewhere that these function symbols are never meant to be attacker symbols (from the signature). We should ideally avoid using `fn_XX` for such functions.

aeyno replied:
Done

LCBH commented on tlspuffin/src/tls/fn_transcript.rs:
```diff
@@ -36,3 +38,35 @@ fn _fn_transcript<T: Transcript>(claim: &T) -> Result<HandshakeHash, FnError> {
+
+pub fn fn_finished_get_client_random(claim: &Finished) -> Result<Random, FnError> {
+    let mut payload = [0; 32];
+    for (idx, v) in claim.client_random.iter().enumerate() {
```
No better way to copy a SmallVec?

aeyno replied:
Done.

LCBH commented on tlspuffin/src/tls/fn_transcript.rs:
```diff
@@ -36,3 +38,35 @@ fn _fn_transcript<T: Transcript>(claim: &T) -> Result<HandshakeHash, FnError> {
+
+pub fn fn_finished_get_server_random(claim: &Finished) -> Result<Random, FnError> {
+    let mut payload = [0; 32];
+    for (idx, v) in claim.server_random.iter().enumerate() {
```
No better way to copy a SmallVec?

aeyno replied:
Done.

LCBH commented on tlspuffin/src/tls/fn_utils.rs:
```diff
@@ -158,6 +166,39 @@ pub fn fn_decrypt_handshake_flight(
     Ok(decrypted_flight)
 }
 
+pub fn fn_decrypt_handshake_flight_with_secret(
```
Add a Rust doc

aeyno replied:
Done.

LCBH commented on tlspuffin/src/claims.rs:
```diff
@@ -60,8 +61,9 @@ impl codec::Codec for TranscriptClientHello {
-#[derive(Debug, Clone)]
-pub struct TranscriptPartialClientHello(pub TlsTranscript);
+#[derive(Debug, Clone, Comparable, PartialEq)]
+pub struct TranscriptPartialClientHello(#[comparable_ignore] pub TlsTranscript);
```
[Compare] Why ignoring this? Shouldn't we compare the length at least? (in TlsTranscript) --> same for other structs in this file

aeyno replied:
All transcripts are ignored since their content must be different.

LCBH commented on tlspuffin/src/claims.rs:
```diff
@@ -169,41 +174,56 @@ pub trait Transcript {
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Comparable, PartialEq)]
 pub struct Finished {
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
This is ignored because it introduces duplicates Finished claims on some PUTs: ignoring this fields makes the oracle deduplicate all adjacent Finished while not ignoring it causes some PUTs to have this claims multiple times with and without `outbound=true`

LCBH commented on tlspuffin/src/claims.rs:
```diff
@@ -169,41 +174,56 @@ pub trait Transcript {
     pub master_secret: SmallVec<[u8; 32]>,
 
     pub chosen_cipher: u16,
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
list of ciphers are different for TLS 1.2, openssl shows 5 (3 TLS 1.3 and two TLS 1.2) while wolfssl shows only the two TLS 1.2 ciphers. So this is an API difference.

LCBH commented on tlspuffin/src/claims.rs:
```diff
@@ -169,41 +174,56 @@ pub trait Transcript {
+    #[comparable_ignore]
     pub available_ciphers: SmallVec<[u16; 20]>,
 
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
It was different for OpenSSL and wolfSSL, but I think you introduced the necessary API to changes sigalgs in https://github.com/tlspuffin/tlspuffin/pull/468. So this could probably be removed when merged.

LCBH commented on tlspuffin/src/claims.rs:
```diff
@@ -169,41 +174,56 @@ pub trait Transcript {
 
+    #[comparable_ignore]
     pub signature_algorithm: i32,
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is this randomized? Different for different PUTs?

aeyno replied:
It was different for OpenSSL and wolfSSL, but I think you introduced the necessary API to changes sigalgs in [!468](https://github.com/tlspuffin/tlspuffin/pull/468). So this could probably be removed when merged.

LCBH commented on tlspuffin/src/claims.rs:
```diff
@@ -251,18 +271,19 @@ pub enum ClaimDataMessage {
     pub origin: AgentType,
     pub protocol_version: TLSVersion,
     pub data: ClaimData,
+    #[comparable_ignore]
```
[Compare] Why ignoring this? Is it because we already test the claim step anyway? But then we could also compare this field?!

aeyno replied:
We ignore the step number because when a claim is present multiple times at multiple steps we can't deduplicate it if they have different step numbers. So we ignore the step number to focus on the succession of claims rather than when they were emitted.

LCBH commented on tlspuffin/src/query.rs:
```diff
@@ -12,6 +12,10 @@ pub enum TlsQueryMatcher {
     Handshake(Option<HandshakeType>),
     ApplicationData,
     Heartbeat,
+    ClientHelloFlight,
```
Add some documentation: For example, this is supposed to match any flight starting with a client hello

aeyno replied:
Done

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
         &TLS_SIGNATURE
     }
+
+    fn differential_fuzzing_blacklist() -> Option<Vec<TypeId>> {
```
Question: None: is this intended? Why did we need a black list then?

aeyno replied:
Blacklist has been removed

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
+        let mut is_client = false;
+        let mut client = AgentName::new();
+
+        for agent in agents {
```
Would be nice to document this

aeyno replied:
Done.

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
+        terms
+    }
+
+    fn differential_fuzzing_claims_blacklist() -> Option<Vec<TypeId>> {
```
[Compare] Why all of these transcripts? Always different?

aeyno replied:
Yes all transcripts are different since they all depend on CH and SH which contains random values.

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
+        ])
+    }
+
+    fn differential_fuzzing_uniformise_put_config(mut trace: Trace<Self>) -> Trace<Self> {
```
Why taking a `mut` on the input? (the signature is weird)

aeyno replied:
done

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
+    }
+
+    fn differential_fuzzing_uniformise_put_config(mut trace: Trace<Self>) -> Trace<Self> {
+        for agent in trace.descriptors.iter_mut() {
```
Document somewhere (ideally on the official doc) for which vendors this works.

LCBH replied:
Also document here why we chose these ciphers.

aeyno replied:
Done.

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
+            *t = Self::differential_fuzzing_uniformise_put_config(t.to_owned());
+        }
+
+        trace
```
Is there a way to test if this has worked (when executing the trace? Is set-cipher capable of detecting failures?)

aeyno replied:
set_cipher and set_groups are tested in `tlspuffin/tests/harness.rs`

LCBH commented on tlspuffin/src/protocol.rs:
```diff
@@ -360,6 +405,110 @@ impl ProtocolTypes for TLSProtocolTypes {
+        }) = diff
+        {
+            // OpenSSL sends an optional EllipticCurves extension when renegociating
+            if diff.contains("EncryptedExtensionsChange([Removed(0, Unknown(UnknownExtensionDesc { typ: EllipticCurves }))])") {
```
This does not resist changes in struct removing or structure modifications. Is there a way to make this test more robust?

aeyno replied:
I changed that to just searching for a difference containing `"UnknownExtensionDesc { typ: EllipticCurves }"`

LCBH commented on tlspuffin/Cargo.toml:
```diff
@@ -340,6 +340,11 @@ enable-guards = []
 # utility functions for testing tlspuffin
 test-utils = ["tempfile"]
 
+ddyf-disable-status = ["puffin/ddyf-disable-status"]
```
please document

aeyno replied:
done

LCBH commented on tools/differential/list_buckets.sh:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
This script needs documentation

aeyno replied:
Done.

LCBH commented on tools/differential/empty_buckets.sh:
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
This script needs documentation

aeyno replied:
Done.

aeyno committed:
[differential fuzzing] adding reproducers for identified RFC violations

aeyno committed:
[differential fuzzing] small refactors and fixes

aeyno commented on puffin/src/codec.rs:
```diff
@@ -417,3 +417,10 @@ impl<T: Codec + VecCodecWoSize> Codec for Vec<T> {
         Some(ret)
     }
 }
+
+pub fn compare_encoding<X: Codec, Y: Codec>(x: &X, y: &Y) -> std::cmp::Ordering {
+    match Codec::get_encoding(x) <= Codec::get_encoding(y) {
+        true => std::cmp::Ordering::Less,
+        false => std::cmp::Ordering::Greater,
+    }
```
done

aeyno committed:
[differential fuzzing] removing differential fuzzing type blacklist

[differential fuzzing] remove blacklist

aeyno commented on puffin/src/trace.rs:
```diff
@@ -24,13 +24,16 @@ use std::marker::PhantomData;
 use std::vec::IntoIter;
 
+use clap::error::Result;
+use comparable::Comparable;
```
Done.

aeyno commented on tools/differential/list_buckets.sh:
```diff
@@ -0,0 +1,22 @@
+  if [ -d "$d" ];
+    then
+    num=$(find $d -maxdepth 1 -type f -regextype posix-egrep -regex '.*\.trace(-[0-9]+)?' | wc -l)
+    total=$((total+num))
```
Done.

aeyno commented on docs/docs/guides/getting-started/differential-fuzzing.md:
```diff
@@ -0,0 +1,169 @@
+
+To triage all the traces in the objectives, we provide a Python library to automate grouping traces.
+
+Start by importing the necessary elements from `./tool/differential/diff-analyzer.py`
```
Done

aeyno commented on docs/docs/guides/getting-started/differential-fuzzing.md:
```diff
@@ -0,0 +1,169 @@
+
+## Requirements
+
+In order to do differential fuzzing, the protocol fuzzer should be configured to minimize the false positives, i.e. all randomized fields (nonces, random, secrets, hash, ciphertext, etc.) should be annotated to be ignored during the comparison step. If the protocol offer some cusomization (e.g. cipher selection), the protocol fuzzer must provide a set of common parameters for all the PUTs to mini ...
```
Done.

aeyno commented on extractable-macro/tests/tests.rs:
```diff
@@ -35,9 +36,37 @@ impl ProtocolTypes for TestProtocolTypes {
+
+    fn differential_fuzzing_blacklist() -> Option<Vec<TypeId>> {
+        todo!()
+    }
```
Blacklists have been removed.

aeyno committed:
[differential fuzzing] small refactoring and adding documentation

[differential fuzzing] add documentation

[differential fuzzing] add documentation

aeyno committed:
[differential fuzzing] update scripts (docs, small fixes) + new RFC violation reproducer trace

aeyno committed:
[differential fuzzing] add check_security_violation in ConfigTrace

aeyno committed:
[differential fuzzing] keep security violations in case of status difference

aeyno commented:
@LCBH can you allow merging this PR ?
